src/formatter/tokenizer.ts

```diff
@@ -0,0 +1,301 @@
+			token += input.charAt(pos++);
+		}
+		if (keywords.includes(token)) {
+			tokenType = "keyword";
+		} else {
+			tokenType = "name";
+		}
+		return token;
+	}
+
+	function readNode() {
+		let token = "";
+		if (!input.charAt(pos).match(/[\&\$\%\^]/)) {
+			return token;
+		}
+		token += input.charAt(pos++);
+		if (input.charAt(pos) === "\"" || input.charAt(pos) === "'") {
+			token += readString();
+		} else {
+			while (input.charAt(pos).match(/[a-z_A-Z0-9\/\%]/)) {
+				token += input.charAt(pos++);
+			}
+		}
+		tokenType = "node";
+		return token;
+	}
+
+	function readNumber() {
+		let token = "";
+		if (input.charAt(pos).match(/[0-9.e\-\_]/)) {
+			token += input.charAt(pos++);
+		}
+		if (input.charAt(pos).match(/[0-9.e\-\_xb]/)) {
+			token += input.charAt(pos++);
+		}
+		if (token === "0x") {
+			while (input.charAt(pos).match(/[a-f0-9.e\-\_]/)) {
+				token += input.charAt(pos++);
+			}
+		} else {
+			while (input.charAt(pos).match(/[0-9.e\-\_]/)) {
+				token += input.charAt(pos++);
+			}
+		}
+		tokenType = "number";
+		return token;
+	}
+
+	function readString() {
+		let token = "";
+		if (input.slice(pos, pos + 3) === "\"\"\"") {
+			token += input.slice(pos, pos + 3);
+			pos += 3;
+			while (pos < input.length && input.slice(pos, pos + 3) !== "\"\"\"") {
+				token += input.charAt(pos++);
+			}
+			token += input.slice(pos, pos + 3);
+			pos += 3;
+			tokenType = "string";
+			return token;
+		}
+		if (input.slice(pos, pos + 3) === "'''") {
+			token += input.slice(pos, pos + 3);
+			pos += 3;
+			while (pos < input.length && input.slice(pos, pos + 3) !== "'''") {
+				token += input.charAt(pos++);
+			}
+			token += input.slice(pos, pos + 3);
+			pos += 3;
+			tokenType = "string";
+			return token;
+		}
+		token += input.charAt(pos++);
+		const quot = token.trim().charAt(0);
+		while (pos < input.length && input.charAt(pos) !== quot) {
+			token += input.charAt(pos++);
+			if (input.charAt(pos - 1) === "\\") {
+				token += input.charAt(pos++);
+			}
+		}
+		token += input.charAt(pos++);
+		tokenType = "string";
+		return token;
+	}
+
+	function between(prev: string, next: string) {
+		if (!prev) return "";
+		if (!next) return "";
+		if (next === "comment") return " ";
+		if (inParamList) {
+			if (prev === "comma") return " ";
+			return "";
+		}
+		if (prev === "symbol") return " ";
+		if (next === "symbol") return " ";
+		if (prev === "comma") return " ";
+		if (next === "comma") return "";
+		if (prev === "bang") return "";
+		if (prev === "keyword") return " ";
+		if (next === "keyword") return " ";
+		if (prev === "parens") return "";
+		if (next === "parens") return "";
+
+		if (prev === "number") return " ";
+		if (prev === "name") return " ";
+		if (prev === "node") return " ";
+		return "";
+	}
+
+	let lastTokenType: string;
+	let token: string;
+	const tokens: string[] = [readWhitespace()];
+	while (pos < input.length) {
+		const char = input.charAt(pos);
+		const nextChar = input.charAt(pos + 1);
```

Comment:
Even though we have the variable `nextChar` we are using `input.charAt(pos + 1)` in a few places, should we replace it with the variable instead for consistency and readability sake?

<img width="622" alt="image" src="https://github.com/godotengine/godot-vscode-plugin/assets/6596836/ce2da3cc-bb1a-42af-b640-fefc3db00cc1">

Reply:
Yeah that's a good catch. 

This entire formatter is basically a placeholder until I have time to finish the real one.